SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift

```diff
@@ -0,0 +1,75 @@
+//===--- ReadOnlyGlobalVariables.swift ------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Marks global `var` variables as `let` if they are never written.
+///
+/// Note that this pass relies on the initialize-static-globals pass which converts lazily
+/// initialized globals to statically initialized globals.
+/// This pass does not mark lazily initialized globals as `let`, because such globals _are_
+/// written: in their initializers.
+///
+let readOnlyGlobalVariablesPass = ModulePass(name: "read-only-global-variables") {
+  (moduleContext: ModulePassContext) in
+
+  var writtenGlobals = Set<GlobalVariable>()
+
+  for f in moduleContext.functions {
+    for inst in f.instructions {
+      if let gAddr = inst as? GlobalAddrInst {
+        if gAddr.addressHasWrites {
+          writtenGlobals.insert(gAddr.global)
+        }
+      }
+    }
+  }
+
+  for g in moduleContext.globalVariables {
+    if !g.isPossiblyUsedExternally,
+       !g.isLet,
+       !writtenGlobals.contains(g) {
+      g.setIsLet(to: true, moduleContext)
+    }
+  }
+}
+
+private extension Value {
+  var addressHasWrites: Bool {
```

Comment:
It's very strange to me that you would hide the functionality of def-use traversal inside the unnecessary abstractions of a computed property within an extension of value. If something is initiating a def-use walk, I *really* want that to be obvious at the point-of-use. It's quite important that people don't accidentally perform repeated def-use walks.

Reply:
Well, a computed property is not more of an abstraction than a method.
But I agree that a property conveys that it's something very simple and doesn't involve any (potential) computational expensive operations.
So in this case it's maybe better to define it as a method.

Reply:
> Well, a computed property is not more of an abstraction than a method.
But I agree that a property conveys that it's something very simple (O(1) complexity) and doesn't involve any (potential) computational expensive operations.

@eeckstein Yes, you said it better. That's the only point I was making. I've seen the same argument made during review of Swift libraries.